File: backend/app/schemas/upload_validation_error.py
"""
Pydantic schemas for upload validation errors.

A validation error is a single bad cell from the Polars validation
layer. One row per bad cell - so a file with 100 type mismatches
produces 100 rows in the upload_validation_errors table.

The endpoint that returns these supports pagination - a single
upload can produce thousands of errors.
"""

from datetime import datetime
from uuid import UUID

from app.schemas.common import ORMBase


class UploadValidationErrorResponse(ORMBase):
    """Response shape for a single validation error row."""

    id: UUID
    upload_id: UUID
    row_number: int
    column_name: str
    error_type: str
    error_message: str
    raw_value: str | None
    created_at: datetime

File: frontend/src/types/upload_validation_error.ts
/**
 * TypeScript types for upload validation errors.
 *
 * Mirrors backend/app/schemas/upload_validation_error.py.
 *
 * One row per bad cell from Polars Layer 1 validation.
 */

/**
 * Validation error categories.
 * Must match the SQL CHECK constraint in
 * sql/08_create_upload_validation_errors_table.sql.
 */
export type ValidationErrorType =
  | 'NOT_NULL'
  | 'UNIQUE'
  | 'TYPE_MISMATCH'
  | 'SCHEMA_MISMATCH'
  | 'ENCODING_ERROR'
  | 'PARSE_ERROR'

export interface UploadValidationError {
  id: string
  upload_id: string
  row_number: number
  column_name: string
  error_type: ValidationErrorType
  error_message: string
  raw_value: string | null
  created_at: string
}